### ISSUE TITLE
Javascript SWIG and Native Abstractions for Node.js

### ISSUE THREAD

**Issue by p2k** [swig/swig#971]
This is a "fork" of #804. I'd like to ask if the general public and the project lead would be interested in adding support for [NaN](https://github.com/nodejs/nan) as another engine/option for the Javascript target on a long term. If so, I'd hereby offer some of my free time to drive this idea forward and convert/implement the necessary code generators one by one.
If you want me to add tests, documentation or examples etc. I can do this as well - it'll just take me some more time and I'm not going to make too many promises on the timeframe.

**Comment by kaspermarstal** [comment]
I can only speak on behalf of myself, and not the general public, but I would be very, very interested in this feature. I am struggling to make node bindings for https://github.com/kaspermarstal/SimpleElastix with the current node driver. I hope your kind offer attracts more attention.

**Comment by p2k** [comment]
Thanks for your interest - I was going to create the fork anyway (as soon as I finish up three projects that consume my free time too).
You can keep this issue subscribed so you get notifications as soon as I make progress.

**Comment by wsfulton** [comment]
Would this make the `-v8` and `-node` options redundant? Or would NaN be used to support these? Sorry for my ignorance, but I also don't understand why we have to have both of these options, All I can find is in SWIG's Javascript documentation
```
The generator for node is essentially delegating to the v8 generator and adds some necessary preprocessor definitions.
```
Seems that NaN negates the need for the preprocessor definitions.

**Comment by p2k** [comment]
Technically NaN is itself a (huge) collection of preprocessor definitions 😉 - but you're right, the `-nan` option would replace `-node` and not delegate to `-v8` but rather be an independent branch. It would make `-node` obsolete, if you like to call it this way, but it would not obsolete `-v8` for all the people that use the V8 in its pure form without Node (e.g. in WebKit).

I wouldn't go that far as to remove the `-node` option straight away. There might be people that don't want the NaN dependency in their Node project. It might just get removed if no one maintains it.

**Comment by p2k** [comment]
Maybe I should elaborate on the `-v8` vs. `-nan` part a bit more. What NaN does is abstracting away almost all V8 calls and types under preprocessor definitions, templates and helper functions. This results in a smaller footprint of code while providing backwards (and forwards) compatibility with Node and V8 changes.

That said, for the NaN option for SWIG/JavaScript one needs to modify nearly all code generating functions, replacing `v8::` with `Nan::` very often (not always) and removing lots of `#if`'s and other checks that NaN does for you. The resulting code will let you compile a Node extension with ease for past, current and possibly future versions of Node.

However, the underlying V8 JavaScript engine wrapped by NaN is not only used in Node. There is Google Chrome, Electron and some other WebKit based browsers that use it and allow writing extensions. For these targets, the generated code with NaN might not compile or the resulting library might not work, because of how Node initializes the library or because of dependencies outside of pure V8 that are provided by Node only.

**Comment by ojwb** [comment]
I doubt it makes sense to keep `-node` around in the hope that people appear to actively maintain it.  They haven't so far, and node has a sufficiently brisk rate of API churn that this is already a significant problem, at least judging by the growing pile of tickets - see https://github.com/swig/swig/labels/Javascript

**Comment by p2k** [comment]
Good point. Given I manage the huge amount of initial work to get NaN going with SWIG, we should be fairly safe from API churns for a while or at least improve the overall situation since there is a chance that NaN fixes the problems by itself without needing a modification from the SWIG side.

**Comment by jdesboeufs** [comment]
I think it is more strategic to directly invest on N-API:
https://medium.com/the-node-js-collection/n-api-next-generation-node-js-apis-for-native-modules-169af5235b06

**Comment by p2k** [comment]
As it turns out, I could not start working on this issue yet. The N-API looks promising and for existing projects, the current SWIG wrapper is fine - so, yes, why not? Can anyone think of a reason NOT to go with the N-API instead of NaN for future SWIG/JS projects?

**Comment by ojwb** [comment]
N-API seems to be self-described as "experimental" currently, and they say "We plan to port N-API to older Node.js LTS lines after it stabilizes in Node.js 8.0" so it seems it doesn't currently actually provide portability across different nodejs versions, and so we'd need to keep all the current code to support older versions around too, and add N-API as a parallel option alongside that.  AIUI with NaN we could just use NaN for all nodejs versions, which seems like it would give a much cleaner result.

But really, if someone is willing to do the work, they probably get to pick which approach to take.

**Comment by p2k** [comment]
I have to look into it, but it could be that the N-API is easier/more straight forward to implement from scratch than NaN; imho we have to keep the "legacy" v8 code anyways - at least for everything that is not Node.js - so the parallel approach is the way to go both for NaN and N-API.

It's hard to find a good argument, there is pretty much equal justification for the two directions to go.

**Comment by murillo128** [comment]
By the way, I think this is important: N-API is no longer experimental in node 10.

https://levelup.gitconnected.com/whats-new-in-node-10-ad360ae55ee4

```
the Node documentation describes N-API as an API for building native Addons. It is independent from the underlying JavaScript runtime (ex V8) and is maintained as part of Node.js itself. This API will be Application Binary Interface (ABI) stable across versions of Node.js. It is intended to insulate Addons from changes in the underlying JavaScript engine and allow modules compiled for one version to run on later versions of Node.js without recompilation.

N-API was experimentally introduced in Node 8 and is going to be to be stable beginning in Node 10. Upgrading between Node versions will no longer cause concern for module breakage. It will also be ported backwards for Node.js v6.x and v8.x compatability.
```

Maybe it is a good time to start working on this and deprecate nan/v8 for node?

**Comment by p2k** [comment]
Cool... looks like I involuntarily sat this one out. Alright then, I'll skip NaN and head over to the N-API; for all other versions below 10, SWIG will continue to offer the legacy method of wrapping that has been there since Node's early days.

**Comment by murillo128** [comment]
btw, is there an issue for N-API support?

**Comment by ojwb** [comment]
There's not one explicitly about it, or else it isn't tagged with "Javascript":

https://github.com/swig/swig/labels/Javascript

But I think #968 covers this - it's really just an approach to solving that.
